man/man1/magalie.1
.TH MAGALIE 1NEMO "25 February 2020"
.SH NAME
magalie \- Boily et al. composite disk-bulge-halo model
.SH SYNOPSIS
\fBmagalie\fP [parameter=value]
.PP
\fBmagalie.exe\fP 
.SH DESCRIPTION
\fBmagalie\fP creates a composite Disk-Bulge-Halo galaxy in 
near equilibrium following the procedure described in
Boily et al.(2000), adapted from Hernquist's (1993) BUILDGAL 
procedure.
This program is merely a NEMO wrapper
to Version 1 of the fortran program \fImagalie.exe\fP.
It assembles all the correct parameters, runs the fortran
namesake program in a temporary rundirectory and converts the
ascii output to a NEMO \fIsnapshot(5NEMO)\fP.
.PP
The galaxy is scaled based on the disk with mass=1 and 
(exponential) scalelength=1. In addition, a
selection of the solar radius (2.43 scale lenghts is the
current default) will fix the rotation speed at the
IAU value of 220 km/s, though the velocity is also normalized
to this value.
.PP
This NEMO frontend only controls a number of parameters, to match
the functionality of \fImkkd95(1NEMO)\fP, however there are quite
a few more options in the original program, e.g.. adding a gas disk,
satellited, individual smoothing lengths. These will be implemented
in a future release. It is easy for the user to manually create
such models those, by setting \fBcleanup=f\fP and editing the
input parameter file \fImagalie.in\fP in the temporary run directory,
and running the \fImake-it\fP shell script. Careful not to change
the number of particles, or modify the corresponding keyword
in this script.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBout=\fP
output snapshot (a rundirectory $out.tmpdir is also created). 
No default.
.TP
\fBndisk=\fP
Number of particles in disk. Set them to 0 to leave out this
component.
Notice these are the first set of particles in the final snapshot.
 [20000]  
.TP
\fBnbulge=\fP
Number of particles in bulge [5000]  
.TP
\fBnhalo=\fP
Number of particles in halo. 
Notice these are the last set of particles in the final snapshot.
[40000]  
.TP
\fBhdisk=\fP
Scaleheight of the disk (in units of scale
.TP
\fBrsolar=\fP
Solaris radius for calibration of Vsolar=220 [2.43]
.TP
\fBQtoomre=\fP
Toomre stability parameter [1.5]
.TP
\fBcdisk=\fP
Cutoff radius of disk [10]
.TP
\fBzdisk=\fP
Height cutoff of disk [2]
.TP
\fBmbulge=\fP
Mass of bulge (in units of disk mass)
.TP
\fBrbulge=\fP
Scalelength of bulge [2]
.TP
\fBcbulge=\fP
Cutoff radius of bulge [10]
.TP
\fBzbulge=\fP
Cutoff height of bulge [1]
.TP
\fBqbulge=\fP
Axis ratio C/A of bulge [1]
.TP
\fBfbulge=\fP
Streaming fraction of bulge [0]
.TP
\fBmhalo=\fP
Mass of halo (in units of disk mass)
.TP
\fBrhalo=\fP
Scalelength of halo [2]
.TP
\fBchalo=\fP
Cutoff radius of halo [20]
.TP
\fBc2halo=\fP
Cutoff radius of halo for ISO (for LH
.TP
\fBqhalo=\fP
Axis ratio C/A of halo [1]
.TP
\fBtype=\fP
Type of halo: LH or ISO [LH]
.TP
\fBseed=\fP
Random seed [0]
.TP
\fBcleanup=t|f\fP
Cleanup runtime working directory?
[Default: t]
.TP
\fBheader=|4|8\fP
Size of the header uses in unformatted fortran data. This used to be 4,
but modern compilers (e.g. gfortran) may very well use 8. By default it
is not specified, and the value determined to be the correct one during
install (UNFIO_HDR_SIZE) is used. This keyword can be used to force
another value, for example if you have compiled magalie with another 
than the default compiler.
.SH INPUTFILE
\fBmagalie\fP calls the actual \fBmagalie.exe\fP by reading an input file
of parameters. Here is an annotated example of such a file (the comments
can be left in when read)
.nf
543216789       ! Random generator seed (ran1)
n               ! Output particles smooth length?
20000           ! Number of disc particles (M_d=1)
3.              ! Mass of disk (reset = 1 for default)
1.              ! Disc scalelength
.1              ! Disc scale height ( = 1/5 length )
2.43            ! Solar radius
1.5             ! Toomre Q parameter @ Sun ( was 3/2 -> close to 1)
.1              ! disc particle smoothing length (set to numerical resolution)
2.              ! z-max (= 10.*scale height)
10.             ! Max disc part radius
n               ! Gas in disc?
0               ! Number of disc gas particles
0.              ! mass of disc gas particles
0.              ! temperature of gas in disc
0.              ! scale height of gas in disc z0gas
0.              ! max z value of gas in disc (= 10 z0gas)
0.              ! max cylindrical radius of gas in disc
0.              ! min    "          "    "   "   "  "
n               ! include gas self-gravity (in disc)? [y/n]
n               ! Add bulge?
0.75            ! Mass of bulge
2.              ! Bulge scale length (hernquist 'a')
y               ! bulge self-gravity?
5000            ! N part in  bulge
10.             ! max radius for bulge particles
.01             ! softening length for particles
n               ! Non-spherical bulge?
.89             ! Value of minor axis ratio ( c/a < 1 )
1.              ! Z-max value for bulge particles
5               ! Number of Simpson integration steps
n               ! Include rotation ? (flip z-momentum component)
.0              ! Fraction of stars with aligned momentum (0<f<1)
n               ! Include a halo?
y               ! halo self-gravity ?
20.             ! max radius of halo
40000           ! Halo particle number
0.01            ! halo particle softening length
LH              ! Halo type (LH or IS)
10.             ! Halo mass
2.              ! Halo core radius  (need two lengths if IS halo)
2.              ! Second length : dummy (LH) or truncation radius (IS)
n               ! Non-spherical halo?
.5              ! aspect ratio (spheroid only)
n               ! Galactic satellite
.1              ! Satellite mass
1.              ! Satellite scale length (LH model only)
5.,0.,0.        ! xyz position of the satellite in Galaxy frame of reference
5.              ! maximum satellite size (compare with sat length scale)
y               ! Include satellite self-gravity ? (yes = live model only)
0               ! number of satellite particles
0.1             ! smoothing length for satelllite particles
b               ! data format for output (a)scii or (b)inary
nbody           ! formatted as LH, nbody
roundbulge      ! name of dataset (note - truncated at cr character)
.fi
.SH PERFORMANCE
The bulk of the CPU is in creating the disk particles, the bulge and halo are 
a much smaller fraction of the cpu. On a 3 GHz Pentium-4 the cpu cost is
about (Ndisk/630) secs for the gnu compiler. On a 1.6 P4 laptop this
was about (Ndisk/420) secs.
.SH BUGS
There have been reported cases where magalie just seems to continue to
compute in an infinite loop.
.PP
Does not work on the linux/intel compiler yet.
.SH SEE ALSO
.nf
mkkd95(1NEMO), tabtos(1NEMO), unfio(1NEMO)
https://ui.adsabs.harvard.edu/abs/2001NewA....6...27B/  (Boily et al. 2001; magalie)
https://ui.adsabs.harvard.edu/abs/1993ApJS...86..389H/abstract (Hernquist 1993)
https://ui.adsabs.harvard.edu/abs/2015MNRAS.452.2367Y/abstract (Yurin & Springel 2015; GalIC)
.SH FILES
.nf
NEMO/src/nbody/init/magalie.c
NEMO/usr/boily/Magalie/
.fi
.SH AUTHOR
.nf
Christian Boily, Pavel Kroupa, Jorge Penarrubia-Garrido (fortran code)
Peter Teuben (NEMO interface) -
.fi
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
dark ages	V1.0 See their NewA paper	Boily et al.
21-Mar-04	V1.1 Created at the Nbody school	PJT
24-mar-04	V1.2 added most primary keywords, at 37,000ft	PJT
25-jan-05	V1.2a fixed bulge mass encoding problem
24-mar-06	V1.2b fixed bulge size encoding problem,updated docs	PJT
19-jul-06	V1.3 add header= for 64 bit compilers	PJT
5-aug-06	merged CVS doc versions		PJT
feb-2020	added refs	PJT
.fi



src/nbody/init/magalie.c
/*  magalie:   make a galaxy (MaGalie; Boily et al. prescription)
 *
 *  <dark-ages> Original fortran code               Boily et al.
 *  21-mar-04   1.1 interface created in rainy Strasbourg           PJT/CB
 *  24-mar-04   1.2 added most primary keywords, at 37,000ft        PJT
 *  23-jan-05   1.2a   fixed bulge mass encoding error     - courtesy J.J.Fleck
 *  24-mar-06   1.2b   fixed bulge radius encoding error   PJT
 *  12-jul-06   1.2c   merged two versions - PJT
 *  19-jul-06   1.3    use header=   PJT

 * TODO: merge in history of how magalie was called...
 */

#include <stdinc.h>
#include <getparam.h>
#include <history.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>

string defv[] = {
  "out=???\n      output snapshot (a rundirectory $out.tmpdir is also created)",
 
  "ndisk=20000\n  Number of particles in disk  (use 0 to skip this component)",
  "nbulge=5000\n  Number of particles in bulge (use 0 to skip this component)",
  "nhalo=40000\n  Number of particles in halo  (use 0 to skip this component)",

  "hdisk=0.2\n    Scaleheight of the disk (in units of scale length)",
  "rsolar=2.43\n  Solaris radius for calibration of Vsolar=220",
  "Qtoomre=1.5\n  Toomre stability parameter",
  "cdisk=10\n     Cutoff radius of disk",
  "zdisk=2\n      Height cutoff of disk",

  "mbulge=0.25\n  Mass of bulge (in units of disk mass)",
  "rbulge=2\n     Scalelength of bulge",
  "cbulge=10\n    Cutoff radius of bulge",
  "zbulge=1\n     Cutoff height of bulge",
  "qbulge=\n      Axis ratio C/A of bulge, in case flattened",
  "fbulge=\n      Streaming fraction of bulge, in case streaming",

  "mhalo=5\n      Mass of halo (in units of disk mass)",
  "rhalo=2\n      Scalelength of halo",
  "chalo=20\n     Cutoff radius of halo",
  "c2halo=20\n    Cutoff radius of halo for ISO (for LH it is c2halo)",
  "qhalo=\n      Axis ratio C/A of halo, in case flattened",
  "type=LH\n      Type of halo: LH or ISO",

  "seed=0\n       Random seed",
  "cleanup=t\n    cleanup run directory after use",
  "header=\n      use an explicit unfio header size of 4 or 8",
  "VERSION=1.3a\n 22-jul-06 PJT",
  NULL,
};

string usage="Boily et al. composite bulge-disk-halo model";

string cvsid="$Id$";

void goto_rundir(string name);
void make_rundir(string name);
void run_program(string cmd);

double getdparamq(string,double);
string getsparamq(string);

void nemo_main(void)
{
  char rundir[256];
  stream datstr, histr;
  string out=getparam("out");
  char hdrkey[32];
  int seed, nbulge, ndisk, nhalo;
  bool Qcleanup = getbparam("cleanup");

  seed =   init_xrandom(getparam("seed")); 
  if (hasvalue("header")) 
    sprintf(hdrkey,"header=%d", getiparam("header"));
  else
    sprintf(hdrkey," ");
  
  nbulge = getiparam("nbulge");
  ndisk  = getiparam("ndisk");
  nhalo  = getiparam("nhalo");
  
  datstr = stropen(out,"w");           /* a dummy write ; should not fail */
  strclose(datstr);
  
  sprintf(rundir,"%s.tmpdir",out);     /* create and change to (tmp) rundir */
  make_rundir(rundir);
  goto_rundir(rundir);
  
  datstr = stropen("magalie.in","w");      /* create input file */
  
  fprintf(datstr,"%d             !! Random generator seed (ran1)\n",seed);
  fprintf(datstr,"n               ! Output particles smooth length?\n");
  fprintf(datstr,"%d             !! Number of disc particles (M_d=1)\n",ndisk);
  fprintf(datstr,"1.              ! Mass of disk (reset = 1 for default)\n");
  fprintf(datstr,"1.              ! Disc scalelength\n");
  fprintf(datstr,"%g             !! Disc scale height ( = 1/5 length )\n",getdparam("hdisk"));
  fprintf(datstr,"%g             !! Solar radius\n",getdparam("rsolar"));
  fprintf(datstr,"%g             !! Toomre Q parameter @ Sun ( was 3/2 -> close to 1)\n",getdparam("Qtoomre"));
  fprintf(datstr,".1              ! disc particle smoothing length (set to numerical resolution)\n");
  fprintf(datstr,"%g             !! z-max (= 10.*scale height)\n",getdparam("zdisk"));
  fprintf(datstr,"%g             !! Max disc part radius\n",getdparam("cdisk"));
  
  fprintf(datstr,"n               ! Gas in disc?\n");
  fprintf(datstr,"0               ! Number of disc gas particles\n");
  fprintf(datstr,"0.              ! mass of disc gas particles\n");
  fprintf(datstr,"0.              ! temperature of gas in disc\n");
  fprintf(datstr,"0.              ! scale height of gas in disc z0gas\n");
  fprintf(datstr,"0.              ! max z value of gas in disc (= 10 z0gas)\n");
  fprintf(datstr,"0.              ! max cylindrical radius of gas in disc\n");
  fprintf(datstr,"0.              ! min    "          "    "   "   "  "\n");
  fprintf(datstr,"n               ! include gas self-gravity (in disc)? [y/n]\n");
  
  fprintf(datstr,"y               ! Add bulge?\n");
  fprintf(datstr,"%g             !! Mass of bulge\n",getdparam("mbulge"));
  fprintf(datstr,"%g             !! Bulge scale length (hernquist 'a')\n",getdparam("rbulge"));
  fprintf(datstr,"y               ! bulge self-gravity?\n");
  fprintf(datstr,"%d             !! N part in  bulge\n",nbulge);
  fprintf(datstr,"%g             ! max radius for bulge particles\n",getdparam("cbulge"));
  fprintf(datstr,".01             ! softening length for particles\n");
  fprintf(datstr,"%s             *! Non-spherical bulge?\n", getsparamq("qbulge"));
  fprintf(datstr,"%g             *! Value of minor axis ratio ( c/a < 1 )\n",getdparamq("qbulge",1.0));
  fprintf(datstr,"%g             !! Z-max value for bulge particles\n",getdparam("zbulge"));
  fprintf(datstr,"5               ! Number of Simpson integration steps\n");
  fprintf(datstr,"%s             *! Include rotation ? (flip z-momentum component)\n",getsparamq("fbulge"));
  fprintf(datstr,"%g             *! Fraction of stars with aligned momentum (0<f<1)\n",getdparamq("fbulge",0.0));
  
  fprintf(datstr,"y               ! Include a halo?\n");
  fprintf(datstr,"y               ! halo self-gravity ?\n");
  fprintf(datstr,"%g             !! max radius of halo\n",getdparam("chalo"));
  fprintf(datstr,"%d             !! Halo particle number\n",nhalo);
  fprintf(datstr,"0.01            ! halo particle softening length\n");
  fprintf(datstr,"%s             !! Halo type (LH or IS)\n",getparam("type"));
  fprintf(datstr,"%g             !! Halo mass\n",getdparam("mhalo"));
  fprintf(datstr,"%g             !! Halo core radius  (need two lengths if IS halo)\n",getdparam("c2halo"));
  fprintf(datstr,"2.              ! Second length : dummy (LH) or truncation radius (IS)\n");
  fprintf(datstr,"%s             *! Non-spherical halo?\n",getsparamq("qhalo"));
  fprintf(datstr,"%g             *! aspect ratio (spheroid only)\n",getdparamq("qhalo",1.0));

  fprintf(datstr,"n               ! Galactic satellite\n");
  fprintf(datstr,".1              ! Satellite mass\n");
  fprintf(datstr,"1.              ! Satellite scale length (LH model only)\n");
  fprintf(datstr,"5.,0.,0.        ! xyz position of the satellite in Galaxy frame of reference\n");
  fprintf(datstr,"5.              ! maximum satellite size (compare with sat length scale)\n");
  fprintf(datstr,"y               ! Include satellite self-gravity ? (yes = live model only)\n");
  fprintf(datstr,"0               ! number of satellite particles\n");
  fprintf(datstr,"0.1             ! smoothing length for satelllite particles\n");

  fprintf(datstr,"b               ! data format for output (a)scii or (b)inary\n");
  fprintf(datstr,"nbody           ! formatted as LH, nbody\n");
  fprintf(datstr,"m.dat           ! name of dataset (note - truncated at cr character)\n");
  strclose(datstr);
   
  histr = stropen("history","w");   /* maintain NEMO style history */
  put_history(histr);
  strclose(histr);
  
  datstr = stropen("make-it","w!"); /* create shell script to be run */
  fprintf(datstr,"#! /bin/sh\n");
  fprintf(datstr,"# created by NEMO's magalie wrapper program VERSION=%s\n",getparam("VERSION"));
  fprintf(datstr,"magalie.exe < magalie.in\n");
  fprintf(datstr,"rm -f ../%s\n",out);
  fprintf(datstr,"unfio in=m.dat block=0 type=f %s | tabtos - ../%s block1=m,x,y,z,vx,vy,vz options=wrap nbody=%d\n",
	  hdrkey, out, ndisk+nbulge+nhalo);
  strclose(datstr);
  run_program("chmod +x make-it; ./make-it > magalie.log 2>&1");   /* run it ! */
  if (Qcleanup) {
    dprintf(1,"Removing the run directory %s.tmpdir\n",out);  
    sprintf(rundir,"cd ..; rm -rf %s.tmpdir",out);  
    run_program(rundir);
  }
}

void goto_rundir(string name)
{
  if (chdir(name))
    error("Cannot change directory to %s",name);
}

void make_rundir(string name)
{
  if (mkdir(name, 0755))
    error("Run directory %s already exists",name);
}

void run_program(string cmd)
{
  if (system(cmd))
    warning("Some error occurred running: %s",cmd);
}

double getdparamq(string key, double def)
{
  return hasvalue(key) ? getdparam(key) : def ;
}

string getsparamq(string key)
{
  return hasvalue(key) ? "y" : "n";
}
